shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -132,19 +140,59 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
   public boolean finishComposingText() {
     boolean result = super.finishComposingText();
 
-    if (Build.VERSION.SDK_INT >= 21) {
-      // Update the keyboard with a reset/empty composing region. Critical on
-      // Samsung keyboards to prevent punctuation duplication.
-      CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
-      builder.setComposingText(-1, "");
-      CursorAnchorInfo anchorInfo = builder.build();
-      mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+    // Apply Samsung hacks. Samsung caches composing region data strangely, causing text
+    // duplication.
+    if (mIsSamsung) {
+      if (Build.VERSION.SDK_INT >= 21) {
+        // Samsung keyboards don't clear the composing region on finishComposingText.
+        // Update the keyboard with a reset/empty composing region. Critical on
+        // Samsung keyboards to prevent punctuation duplication.
+        CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
+        builder.setComposingText(-1, "");
```

Comment:
optional/subjective nit here and throughout: It makes API calls like this easier to read if the primitives are annotated so you can tell what they're supposed to be controlling.

```suggestion
        builder.setComposingText(/*composingTextStart=*/-1, /*composingText=*/"");
```

Reply:
good idea, much clearer what this is doing!